lib/galaxy/managers/users.py

```diff
@@ -109,10 +111,108 @@ def create(self, email=None, username=None, password=None, **kwargs):
         return user
 
     def delete(self, user):
+        """Mark the given user deleted."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to delete users.')
+        # Check first if it is already deleted. If so, say so and do nothing.
+        # Unsure if we should add username to the log.debug or if that breaks the GDPR support
+        if user.deleted:
+            log.debug("User already deleted, but setting delete flag just in case.")
+        if user.purged:
+            log.debug("Cannot delete purged user, but setting delete flag just in case.")
         user.deleted = True
         self.session().add(user)
         self.session().flush()
 
+    def undelete(self, user):
+        """Remove the deleted flag for the given user."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to undelete users.')
+        if user.purged:
+            raise exceptions.ItemDeletionException('Purged user cannot be undeleted.')
+        if not user.deleted:
+            log.debug('User \'%s\' has not been deleted, so it cannot be undeleted.' % user.email)
+        user.deleted = False
+        self.session().add(user)
+        self.session().flush()
+
+    def purge(self, user):
+        """Purge the given user. They must have the deleted flag already."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to delete or purge users.')
+        if not user.deleted:
+            raise exceptions.MessageException('User \'%s\' has not been deleted, so they cannot be purged.' % user.email)
```

Comment:
I'd remove lines 143-144, purging without deleting first is fine by me.

Reply:
I think it is a useful extra safety layer before a destructive function is called. @cganote was it meant like this?

Reply:
Yep I was hoping to have a little biut of a safety net

Reply:
I don't think this check belongs to the manager.

Reply:
I think it is better here than in the API controller though. Where would you put it?

Reply:
In the UI (`lib/galaxy/webapps/galaxy/controllers/admin.py` ?), I don't think it's the API business to ask users for confirmation or similar.

Reply:
History purge doesn't have this check in either the manager or the API.